tests/testthat/test-utils.R

```diff
@@ -43,6 +43,73 @@ test_that("getSelector", {
   expect_equal(getSelector("mykey", "%", exact = FALSE), " AND mykey LIKE '%'")
 })
 
+test_that("errors are rethrown informatively (#643, #788)", {
+  skip_if_not(has_unixodbc())
+
+  expect_snapshot(error = TRUE, dbConnect(odbc(), dsn = "does_not_exist_db"))
+
+  con <- test_con("SQLITE")
+  expect_snapshot(error = TRUE, dbExecute(con, "SELECT * FROM boopbopbopbeep"))
+})
+
+test_that("rethrow_database_error() errors well when parse_database_error() fails", {
+  expect_snapshot(
+    error = TRUE,
+    rethrow_database_error("boop", call = NULL)
+  )
+})
+
+test_that("parse_database_error() works with messages from the wild", {
+  expect_snapshot(
+    error = TRUE,
+    rethrow_database_error(
+      "nanodbc/nanodbc.cpp:1135: 00000
+       [unixODBC][Driver Manager]Data source name not found and no default driver specified",
+      call = NULL
+    )
+  )
```

Comment:
I'd pull the message out of the snapshot avoid some of cruft:

```suggestion
  msg <- "nanodbc/nanodbc.cpp:1135: 00000
       [unixODBC][Driver Manager]Data source name not found and no default driver specified"
  expect_snapshot(rethrow_database_error(msg, call = NULL), error = TRUE)
  )
```